File: pytools/examples/java/client/protocol/ReqTestPhp.java
package protocol;


import network.Packet;
import network.PacketUtil;

import java.util.List;
import java.util.ArrayList;


public class ReqTestPhp
{
	private long u64;
	private String strxx;
	private MsgTestPhp msg_req;
	private byte msg_opt_flag;
	private MsgTestPhp msg_opt;
	private List<Msgtestphp> msg_rep = new ArrayList<Msgtestphp>();


	public byte[] encode()
	{
		Packet packet = new Packet();
		packet.writeLong(u64);
		packet.writeString(strxx);
		packet.writeBytes(msg_req.getBytes());
		packet.writeByte(msg_opt_flag);
		if (msg_opt_flag == 1)
		{
			packet.writeBytes(msg_opt.getBytes());
		}
		int msg_rep_count = msg_rep.size();
		packet.writeShort((short)msg_rep_count);
		for (int i = 0; i < msg_rep_count; i++)
		{
			packet.writeBytes(msg_rep.get(i).getBytes());
		}
		return packet.encode((short)Msg.P_REQ_TEST_PHP);
	}



	public void setU64(long u64)
	{
		this.u64 = u64;
	}

	public void setStrxx(String strxx)
	{
		this.strxx = strxx;
	}

	public void setMsgReq(MsgTestPhp msg_req)
	{
		this.msg_req = msg_req;
	}

	public void setMsgOpt(MsgTestPhp msg_opt)
	{
		this.msg_opt_flag = (byte)1;
		this.msg_opt = msg_opt;
	}

	public void setMsgRep(List<Msgtestphp> msg_rep)
	{
		this.msg_rep = msg_rep;
	}

}

File: pytools/examples/cpp/client/protocol/ReqTestPhp.cpp
#ifndef _REQ_TEST_PHP_
#define _REQ_TEST_PHP_

#include <list>

#include <pb_type.hpp>
#include <Packet.hpp>

#include <Msg.hpp>

#include <MsgTestPhp.cpp>
#include <MsgTestPhp.cpp>
#include <MsgTestPhp.cpp>


class ReqTestPhp
{
private:
	U64 m_u64;
	String m_strxx;
	MsgTestPhp* m_msg_req;
	U8 m_msg_opt_flag;
	MsgTestPhp* m_msg_opt;
	list<MsgTestPhp*> m_msg_rep;


public:
	Packet Encode()
	{
		Packet packet;
		packet.WriteU64(m_u64);
		packet.WriteString(m_strxx);
		U32 msg_req_buff_len = 0;
		BYTE* msg_req_buff = m_msg_req->GetBytes(&msg_req_buff_len);
		packet.WriteBytes(msg_req_buff, msg_req_buff_len);
		packet.WriteU8(m_msg_opt_flag);
		if (m_msg_opt_flag == 1)
		{
			U32 msg_opt_buff_len = 0;
			BYTE* msg_opt_buff = m_msg_opt->GetBytes(&msg_opt_buff_len);
			packet.WriteBytes(msg_opt_buff, msg_opt_buff_len);
		}
		U16 msg_rep_count = (U16)m_msg_rep.size();
		packet.WriteU16(msg_rep_count);
		for (list<MsgTestPhp*>::iterator i = m_msg_rep.begin(); i != m_msg_rep.end(); i++)
		{
			U32 msg_rep_buff_len = 0;
			MsgTestPhp* xx = (MsgTestPhp*)*i;
			BYTE* msg_rep_buff = xx->GetBytes(&msg_rep_buff_len);
			packet.WriteBytes(msg_rep_buff, msg_rep_buff_len);
		}
		packet.Encode(Msg::P_REQ_TEST_PHP);
		return packet;
	}


	void SetU64(U64 u64)
	{
		m_u64 = u64;
	}

	void SetStrxx(String strxx)
	{
		m_strxx = strxx;
	}

	void SetMsgReq(MsgTestPhp* msg_req)
	{
		m_msg_req = msg_req;
	}

	void SetMsgOpt(MsgTestPhp* msg_opt)
	{
		m_msg_opt_flag = 1;
		m_msg_opt = msg_opt;
	}

	void SetMsgRep(list<MsgTestPhp*> msg_rep)
	{
		m_msg_rep = msg_rep;
	}

};

#endif
